app_flutter/lib/widgets/task_grid.dart

```diff
@@ -47,12 +49,367 @@ class TaskGridContainer extends StatelessWidget {
+      return false;
+    }
+    return true;
+  }
+
+  bool matchesTask(QualifiedTask qualifiedTask) {
+    if (taskFilter != null && !taskFilter.hasMatch(qualifiedTask.task)) {
+      return false;
+    }
+    bool stageFlag;
+    switch (qualifiedTask.stage) {
+      case StageName.devicelab:
+        stageFlag = showAndroid;
+        break;
+      case StageName.devicelabIOs:
+        stageFlag = showIos;
+        break;
+      case StageName.devicelabWin:
+        stageFlag = showWindows;
+        break;
+      case StageName.cirrus:
+        stageFlag = showCirrus;
+        break;
+      case StageName.luci:
+        stageFlag = showLuci;
+        break;
+      default:
+        return false;
+    }
+    return stageFlag;
+  }
+
+  Map<String,String> toMap() => <String,String>{
+    if (taskFilter != defaultFilter.taskFilter)
+      'taskFilter': taskFilter.pattern,
+    if (authorFilter != defaultFilter.authorFilter)
+      'authorFilter': authorFilter.pattern,
+    if (messageFilter != defaultFilter.messageFilter)
+      'messageFilter': messageFilter.pattern,
+    if (showAndroid != defaultFilter.showAndroid)
+      'showAndroid': showAndroid.toString(),
+    if (showIos != defaultFilter.showIos)
+      'showIos': showIos.toString(),
+    if (showWindows != defaultFilter.showWindows)
+      'showWindows': showWindows.toString(),
+    if (showCirrus != defaultFilter.showCirrus)
+      'showCirrus': showCirrus.toString(),
+    if (showLuci != defaultFilter.showLuci)
+      'showLuci': showLuci.toString(),
+  };
+
+  String _stringFor(String label, Object value) => value == null ? '' : '$label: $value,';
+
+  @override
+  String toString() {
+    return 'TaskGridFilter('
+        '${_stringFor('taskFilter',    taskFilter?.pattern)}'
+        '${_stringFor('authorFilter',  authorFilter?.pattern)}'
+        '${_stringFor('messageFilter', messageFilter?.pattern)}'
+        '${_stringFor('showAndroid',   showAndroid)}'
+        '${_stringFor('showIos',       showIos)}'
+        '${_stringFor('showWindows',   showWindows)}'
+        '${_stringFor('showCirrus',    showCirrus)}'
+        '${_stringFor('showLuci',      showLuci)}'
+        ')';
+  }
+
+  @override
+  int get hashCode {
+    return hashValues(
+      taskFilter,
+      authorFilter,
+      messageFilter,
+      showAndroid,
+      showIos,
+      showWindows,
+      showCirrus,
+      showLuci,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is TaskGridFilter &&
+        taskFilter == other.taskFilter &&
+        authorFilter == other.authorFilter &&
+        messageFilter == other.messageFilter &&
+        showAndroid == other.showAndroid &&
+        showIos == other.showIos &&
+        showWindows == other.showWindows &&
+        showCirrus == other.showCirrus &&
+        showLuci == other.showLuci
+    ;
+  }
+}
+
+class TaskGridFilterWidget extends StatefulWidget {
```

Comment:
The filter logic should be moved into `logic/task_grid_filter.dart`

Reply:
Done, along with creating filter_property_sheet.dart in the widgets directory as well.